libs/Microsoft.MixedReality.WebRTC/LocalVideoTrack.cs

```diff
@@ -259,30 +296,32 @@ public override void Dispose()
             Debug.Assert(Transceiver == null);
 
             // Unregister interop callbacks
-            if (_selfHandle != IntPtr.Zero)
-            {
-                LocalVideoTrackInterop.LocalVideoTrack_RegisterI420AFrameCallback(_nativeHandle, null, IntPtr.Zero);
-                LocalVideoTrackInterop.LocalVideoTrack_RegisterArgb32FrameCallback(_nativeHandle, null, IntPtr.Zero);
-                Utils.ReleaseWrapperRef(_selfHandle);
-                _selfHandle = IntPtr.Zero;
-                _interopCallbackArgs = null;
-            }
+            _videoFrameReady = null;
+            _argb32VideoFrameReady = null;
+            Utils.ReleaseWrapperRef(_selfHandle);
+            _selfHandle = IntPtr.Zero;
 
             // Destroy the native object. This may be delayed if a P/Invoke callback is underway,
             // but will be handled at some point anyway, even if the managed instance is gone.
             _nativeHandle.Dispose();
         }
 
-        internal void OnI420AFrameReady(I420AVideoFrame frame)
+        void VideoTrackSourceInterop.IVideoSource.OnI420AFrameReady(I420AVideoFrame frame)
         {
             MainEventSource.Log.I420ALocalVideoFrameReady(frame.width, frame.height);
-            I420AVideoFrameReady?.Invoke(frame);
+            lock(_videoFrameReadyLock)
```

Comment:
Did you check the lock was not adding any performance overhead? I don't think so, but since this is the critical path...

Reply:
I didn't, I would expect it to be in the order of 10-100 ns if uncontested, so irrelevant.

I am inclined to take out this lock anyway, it can cause deadlocks and the only side effect of removing it is that a handler may be called once (at most) after it has been removed. This is an issue with any free-threaded C# event, so IMHO users should be aware of this and handle any necessary synchronization (we should point it out in the comments ofc as it's not obvious). Otherwise for consistency we should implement this pattern in all our free-threaded events which would be impractical.

Reply:
Alright, that makes sense. If there's not much benefit then let's keep it simple and remove it.